File: showcase/starters/langgraph-typescript/agent/tool-rendering-reasoning-chain.ts
/**
 * Tool Rendering (Reasoning Chain) — TypeScript port of
 * tool_rendering_reasoning_chain_agent.py.
 *
 * Minimal ReAct agent with tools: reasoning tokens plus sequential tool
 * calls for the tool-rendering-reasoning-chain cell.
 */

import { z } from "zod";
import { tool } from "@langchain/core/tools";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { ChatOpenAI } from "@langchain/openai";

const SYSTEM_PROMPT =
  "You are a travel & lifestyle concierge. When a user asks a question, " +
  "reason step-by-step and call 2+ tools in succession when relevant.";

const getWeather = tool(
  async ({ location }) => ({
    city: location,
    temperature: 68,
    humidity: 55,
    wind_speed: 10,
    conditions: "Sunny",
  }),
  {
    name: "get_weather",
    description: "Get the current weather for a given location.",
    schema: z.object({
      location: z.string().describe("City name"),
    }),
  },
);

const searchFlights = tool(
  async ({ origin, destination }) => ({
    origin,
    destination,
    flights: [
      {
        airline: "United",
        flight: "UA231",
        depart: "08:15",
        arrive: "16:45",
        price_usd: 348,
      },
      {
        airline: "Delta",
        flight: "DL412",
        depart: "11:20",
        arrive: "19:55",
        price_usd: 312,
      },
      {
        airline: "JetBlue",
        flight: "B6722",
        depart: "17:05",
        arrive: "01:30",
        price_usd: 289,
      },
    ],
  }),
  {
    name: "search_flights",
    description:
      "Search mock flights from an origin airport to a destination airport.",
    schema: z.object({
      origin: z.string().describe("Origin airport code"),
      destination: z.string().describe("Destination airport code"),
    }),
  },
);

const getStockPrice = tool(
  async ({ ticker }) => {
    const randInt = (lo: number, hi: number) =>
      Math.floor(Math.random() * (hi - lo + 1)) + lo;
    const sign = Math.random() < 0.5 ? -1 : 1;
    return {
      ticker: ticker.toUpperCase(),
      price_usd:
        Math.round((100 + randInt(0, 400) + randInt(0, 99) / 100) * 100) / 100,
      change_pct: Math.round(sign * (randInt(0, 300) / 100) * 100) / 100,
    };
  },
  {
    name: "get_stock_price",
    description: "Get a mock current price for a stock ticker.",
    schema: z.object({
      ticker: z.string().describe("Stock ticker symbol"),
    }),
  },
);

const rollDice = tool(
  async ({ sides }) => {
    const n = sides ?? 6;
    const max = Math.max(2, n);
    return { sides: n, result: Math.floor(Math.random() * max) + 1 };
  },
  {
    name: "roll_dice",
    description: "Roll a single die with the given number of sides.",
    schema: z.object({
      sides: z
        .number()
        .int()
        .optional()
        .describe("Number of sides on the die (default 6)"),
    }),
  },
);

// Route through a reasoning-capable model via the Responses API so the
// chain of thought streams as AG-UI `ReasoningMessage` events alongside
// the tool calls. Falls back to gpt-4o-mini (no reasoning stream) if
// `OPENAI_REASONING_MODEL` is unset.
const REASONING_MODEL = process.env.OPENAI_REASONING_MODEL ?? "gpt-5-mini";

const model = new ChatOpenAI({
  model: REASONING_MODEL,
  useResponsesApi: true,
  reasoning: { effort: "low", summary: "auto" },
});

export const graph = createReactAgent({
  llm: model,
  tools: [getWeather, searchFlights, getStockPrice, rollDice],
  prompt: SYSTEM_PROMPT,
});

File: showcase/integrations/ag2/src/agents/tool_rendering_reasoning_chain.py
"""AG2 agent for the Tool Rendering (Reasoning Chain) demo.

A travel & lifestyle concierge that chains 2+ tool calls in succession
when relevant. The frontend wires renderers for `get_weather` and
`search_flights` plus a custom catch-all for the rest.

Note: ag2 1.0's AGUIStream maps model reasoning deltas to AG-UI
REASONING_MESSAGE_* events natively, so the reasoning slot can render
streaming "thinking…" text when the underlying model emits reasoning.
The cell still exercises the full tool-rendering chain either way.
"""

from __future__ import annotations

import json
from random import choice, randint
from typing import Annotated

from ag2 import Agent
from ag2.ag_ui import AGUIStream
from ag2.config import OpenAIConfig
from fastapi import FastAPI
from pydantic import Field


async def get_weather(
    location: Annotated[str, Field(description="City or place to look up the weather for")],
) -> dict:
    """Get the current weather for a given location."""
    return {
        "city": location,
        "temperature": 68,
        "humidity": 55,
        "wind_speed": 10,
        "conditions": "Sunny",
    }


async def search_flights(
    origin: Annotated[str, Field(description="Origin airport code, e.g. 'SFO'")],
    destination: Annotated[str, Field(description="Destination airport code, e.g. 'JFK'")],
) -> str:
    """Search mock flights from an origin airport to a destination."""
    payload = {
        "origin": origin,
        "destination": destination,
        "flights": [
            {
                "airline": "United",
                "flight": "UA231",
                "depart": "08:15",
                "arrive": "16:45",
                "price_usd": 348,
            },
            {
                "airline": "Delta",
                "flight": "DL412",
                "depart": "11:20",
                "arrive": "19:55",
                "price_usd": 312,
            },
            {
                "airline": "JetBlue",
                "flight": "B6722",
                "depart": "17:05",
                "arrive": "01:30",
                "price_usd": 289,
            },
        ],
    }
    return json.dumps(payload)


async def get_stock_price(
    ticker: Annotated[str, Field(description="Stock ticker symbol (e.g. AAPL, TSLA, MSFT)")],
) -> dict:
    """Get a mock current price for a stock ticker."""
    return {
        "ticker": ticker.upper(),
        "price_usd": round(100 + randint(0, 400) + randint(0, 99) / 100, 2),
        "change_pct": round(choice([-1, 1]) * (randint(0, 300) / 100), 2),
    }


async def roll_dice(
    sides: Annotated[int, Field(description="Number of sides on the die (default 6)")] = 6,
) -> dict:
    """Roll a single die with the given number of sides."""
    return {"sides": sides, "result": randint(1, max(2, sides))}


SYSTEM_PROMPT = (
    "You are a travel & lifestyle concierge. When a user asks a question, "
    "reason step-by-step and call 2+ tools in succession when relevant. "
    "For weather + travel questions, call get_weather then search_flights. "
    "Keep the final summary to one short sentence."
)


agent = Agent(
    name="tool_rendering_reasoning_chain_assistant",
    prompt=SYSTEM_PROMPT,
    config=OpenAIConfig(model="gpt-4o-mini", streaming=True),
    tools=[get_weather, search_flights, get_stock_price, roll_dice],
)

stream = AGUIStream(agent)
tool_rendering_reasoning_chain_app = FastAPI()
tool_rendering_reasoning_chain_app.mount("", stream.build_asgi())
